### PR TITLE
Changed MarkingDelegate scanClass to reuse ClassIteratorClassSlots

### PR DESCRIPTION
**Pull request by OussamaSaoudi**
Change added to o reduce code redundancy and simplify code.

Currently disables ITable optimization which only looks at ITable
values for boolean arrays since all arrays share the same ITable.

Signed-off-by: Oussama Saoudi <oussama.saoudi@ibm.com>

### CODE REVIEW

**Comment by OussamaSaoudi** [comment]
@dmitripivkine

File: runtime/gc_glue_java/MarkingDelegate.cpp [hunk 1]
```diff
@@ -386,47 +386,10 @@ MM_MarkingDelegate::scanClass(MM_EnvironmentBase *env, J9Class *clazz)
...
-		if (!GC_ClassModel::usesSharedITable(javaVM, clazz)) {
-			J9ITable *clazzITable = (J9ITable *)clazz->iTable;
-			J9ITable *endITable = (0 != classDepth) ? (J9ITable *)(clazz->superclasses[classDepth - 1]->iTable) : NULL;
-			while (clazzITable != endITable) {
-				_markingScheme->markObject(env, clazzITable->interfaceClass->classObject);
-				clazzITable = clazzITable->next;
-			}
+		GC_ClassIteratorClassSlots classSlotIterator(clazz);
+		J9Class **slotPtr;
+		while((NULL != (slotPtr = classSlotIterator.nextSlot())) && (NULL != *slotPtr)) {
```
Comment (dmitripivkine) [inline runtime/gc_glue_java/MarkingDelegate.cpp:391]: This is not correct. This code stops iteration on first classSlot contain NULL.
You need to separate iterator end_of_iteration condition `(NULL != slotPtr)` and actual handling condition `(NULL != *slotPtr)`:
```
    while (NULL != (slotPtr = classSlotIterator.nextSlot())) {
        if (NULL != *slotPtr) {
            _markingScheme->markObject(env, (*slotPtr)->classObject);
        }
    }
```

File: runtime/gc_glue_java/MarkingDelegate.cpp [hunk 2]
```diff
@@ -386,46 +386,11 @@ MM_MarkingDelegate::scanClass(MM_EnvironmentBase *env, J9Class *clazz)
...
-		J9JavaVM *javaVM = (J9JavaVM *)env->getLanguageVM();
-		if (!GC_ClassModel::usesSharedITable(javaVM, clazz)) {
-			J9ITable *clazzITable = (J9ITable *)clazz->iTable;
-			J9ITable *endITable = (0 != classDepth) ? (J9ITable *)(clazz->superclasses[classDepth - 1]->iTable) : NULL;
-			while (clazzITable != endITable) {
-				_markingScheme->markObject(env, clazzITable->interfaceClass->classObject);
-				clazzITable = clazzITable->next;
+		GC_ClassIteratorClassSlots classSlotIterator(clazz);
+		J9Class **slotPtr;
+		while(NULL != (slotPtr = classSlotIterator.nextSlot())){
```
Comment (dmitripivkine) [inline runtime/gc_glue_java/MarkingDelegate.cpp:391]:
```suggestion
		while (NULL != (slotPtr = classSlotIterator.nextSlot())){
```

**Review by dmitripivkine — APPROVED** [review]

**Comment by dmitripivkine** [comment]
Jenkins compile all jdk11

**Comment by dmitripivkine** [comment]
Jenkins test sanity,extended xLinux jdk11

File: runtime/gc_glue_java/MarkingDelegate.cpp [hunk 3]
```diff
@@ -386,46 +386,11 @@ MM_MarkingDelegate::scanClass(MM_EnvironmentBase *env, J9Class *clazz)
...
-
-		/* ITable */
-		J9JavaVM *javaVM = (J9JavaVM *)env->getLanguageVM();
-		if (!GC_ClassModel::usesSharedITable(javaVM, clazz)) {
-			J9ITable *clazzITable = (J9ITable *)clazz->iTable;
-			J9ITable *endITable = (0 != classDepth) ? (J9ITable *)(clazz->superclasses[classDepth - 1]->iTable) : NULL;
-			while (clazzITable != endITable) {
-				_markingScheme->markObject(env, clazzITable->interfaceClass->classObject);
-				clazzITable = clazzITable->next;
+		GC_ClassIteratorClassSlots classSlotIterator(clazz);
```
Comment (dmitripivkine) [inline runtime/gc_glue_java/MarkingDelegate.cpp:389]: You forgot to update signature of the call, now it should be `(javaVM, clazz)`

**Comment by dmitripivkine** [comment]
Jenkins compile all jdk11

**Comment by dmitripivkine** [comment]
Jenkins test sanity,extended xLinux jdk11

**Comment by dmitripivkine** [comment]
Hang in jdmpview is not related to change, merging

**Pull request merged**